### PR TITLE
[DRAFT] Assembler

### PR DESCRIPTION
**Pull request by clooudch**
This PR includes the following major changes:
- Instructions have now a opcode length and return an array of halfwords
- PC is incremented in the processor BEFORE the instructions are executed (because of pipelining). Otherwise the PC offset is calculated wrong and the instructions need the possibility to overwrite the next PC (also MOVS and LDR instructions)
- Instructions can use label offsets
  - to use label offsets set 'needsLabels' to true
  - the assembler will then encode the instruction twice
  - the first time with labels = undefined (the instruction should just use a placeholder value)
  - the second time with correct label offsets (the instruction should use the specified labels and throw if not exists)
- assembler now produces two segments on for code and data
  - segments are loaded and re-loaded on reset by the ct board
- data instructions can depend on labels and will automatically be updated during re-location in the linker
- pseudo instructions will create a literal pool at the end of the code section
  - parser now allows '=' as a option character as well
  - if the value in the literal pool also depends on a symbol, it will be updated as well

The PR also includes two basic LDR and B instructions because they are used for manual testing the assembler behavior. Also some bugs in the LDR instructions were fixed and the parser now enforces upper case for the instruction name.

The following demo assembly is working as expected:

```
CONSTANT EQU 0x123
AREA |.data|, DATA, READWRITE 
    DCD CONSTANT ; the value of CONSTANT symbol
    DCD variable1 ; the address of variable1
    DCD label1 ; the address of label1
variable1
    DCD 0x234 ; a constant value 
AREA |.text|, CODE, READONLY
    LDR R1, label1
    B label1
    MOVS R1, #1 ; skipped
    MOVS R1, #1 ; skipped
    MOVS R1, #1 ; skipped
label1
    LDR R3, =0x456 ; load constant from literal pool
    LDR R4, =CONSTANT ; load symbolic constant from
                      ;  literal pool
    LDR R5, =variable1 ; load address of variable
                       ; from literal pool 
    LDR R6, [R7, #0] ; load variable1 from address
```

Known issues are:
- B instruction is currently unable to handle negative pc increments
- instruction 'LDR R1, [ R2 ]' is not working (there exists already a bug for this)
- there are missing 'LDR' instructions with labels ('LDR R1, label' is implemented as a demo but the other variants like 'LDRB R1, label' are still missing
-  we could think about adding support for 'MOVS R1, SYMBOL', but this would require that the instruction has access to the label values (not only offsets) as well.

### CODE REVIEW

File: src/board/index.ts [hunk 1]
```diff
@@ -15,46 +16,67 @@ class Board {
...
       this.leds
     ])
     this.processor = new Processor(this.registers, this.memory, InstructionSet)
+    this.processor.on('reset', () => this.reloadExecutable())
+  }
+
+  /**
+   * Gets the executable which is currently loaded to the board
+   *
+   * @returns
```
Comment (clooudch) [inline src/board/index.ts:47]: Missing documentation

**Comment by ni507** [comment]
@clooudch I fixed the known issue (B instruction is currently unable to handle negative pc increments) in the branch https://github.com/ZHAW-FS22-PM4/Virtual-CT-Board/tree/vcb-13-unconditional-jump

**Comment by clooudch** [comment]
Maybe it would be better if we could avoid creating a dedicated type just for this one scenario with 11 bits. Can't we make the binary base type non-abstract and just instantiate it using the correct settings.

Imagine having:
```typescript
export class BinaryValue {
  public readonly value: number
  public readonly length: number
  public readonly maxValue: number
  constructor(value: number, length: number) {
    this.value = value
    this.length= length
    this.maxValue = Math.pow(2, length)
  }
}

export class Word extends BinaryValue {
  constructor(value: number) {
    super(value, 32)
  }
}

function demo() {
  const imm11 = new BinaryValue(-20, 11)
}
```

File: src/assembler/elf/utils.ts [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+ */
+export function getSymbolsFromSection(file: IELF, section: string): ISymbol[] {
+  return Object.values(file.symbols).filter((x) => x.section === section)
+}
+
+/**
+ * Gets all the symbols of a specific type.
+ *
+ * @param file the file
+ * @param file the filearam type the type of symbol
```
Comment (ni507) [inline src/assembler/elf/utils.ts:105]: typo 'filearam' and wrong param name

File: src/instruction/interfaces.ts [hunk 3]
```diff
@@ -45,16 +59,21 @@ export interface IInstructionExecutor {
...
 
+  /**
+   * The length of the instruction in halfwords.
+   */
+  opcodeLength: number
+
   /**
    * Executes the instruction represented by the specified opcode.
    *
    * @param opcode the opcode of the instruction to be executed
```
Comment (ni507) [inline src/instruction/interfaces.ts:70]: Missing param docs for `registers` and `memory`

File: src/instruction/interfaces.ts [hunk 4]
```diff
@@ -1,12 +1,15 @@
...
 
 /**
  * Represents a list of label offsets.
  */
 export interface ILabelOffsets {
-  [name: string]: Halfword
+  /**
+   * The offset to the label in range -2147483648 to 2147483647 (signed 32 bit).
+   */
+  [name: string]: Word
```
Comment (ni507) [inline src/instruction/interfaces.ts:12]: The label offsets for the jump instructions are never greater than a halfword. Is this change to the type `Word`required for pseudo instructions?

**Review by ni507 — DISMISSED** [review]
In general, things are looking good. However, with such big changes it is difficult to look at everything in detail, so I rather just looked at it a bit roughly.

File: src/instruction/interfaces.ts (hunk 4 above)
Comment (ni507) [reply src/instruction/interfaces.ts:12]: Nevermind. The offset for the "BL" instruction is larger than a halfword.

File: src/board/processor.ts [hunk 5]
```diff
@@ -102,16 +105,18 @@ export class Processor extends EventEmitter<ProcessorEvents> {
...
       this.emit('endOfCode')
       return
     }
-    this.instructions
-      .getExecutor(opcode)
-      .executeInstruction(opcode, this.registers, this.memory)
-    this.registers.writeRegister(Register.PC, pc.add(2))
+    const executor = this.instructions.getExecutor(opcode[0])
+    for (let i = 1; i < executor.opcodeLength; i++) {
+      opcode.push(this.memory.readHalfword(pc.add(i)))
```
Comment (ni507) [inline src/board/processor.ts:116]: Little bug. PC register should be incremented by `i * 2` to read the correct halfword from the memory.
Fixed this in the branch vcb-13-unconditional-jump.

File: src/assembler/encoder.ts [hunk 6]
```diff
@@ -1,79 +1,289 @@
...
+  )
+}
+
+/**
+ * Writes a data instruction to the file writer.
+ *
+ * @param writer the file writer
+ * @param instruction the data instruction to write
+ */
+function writeDataInsruction(
```
Comment (m-dot-wigger) [inline src/assembler/encoder.ts:205]: typo Instruction

**Review by m-dot-wigger — CHANGES_REQUESTED** [review]
as far as I can see LDR R3, label is only possible for LDR and not for LDRB or LDRH 
grafik


 codecoverage is not best, but will increase with branches depending on this:
grafik


I don't know if it helps readability if on 16 bit instruction opcode array is set to `opcode = opcode[0]` in the onExecuteInstruction

**Comment by m-dot-wigger** [comment]
This PR will be merged into master for VCB-93 Data sections, right?
So I can merge this into my bug branch to resolve merge conflicts

**Comment by deflox** [comment]
Regarding "parser now enforces upper case for the instruction name.". Just as a sidenote: The real CT board allows lowercase instruction names.

**Comment by clooudch** [comment]
> Regarding "parser now enforces upper case for the instruction name.". Just as a sidenote: The real CT board allows lowercase instruction names.

It does still support lowercase instructions but they are now converted to upper case during parsing. So, that the 'canEncodeInstruction' always receives upper case.

**Review by m-dot-wigger — APPROVED** [review]
All looks good now can be merged.

Also as deputy for Silvan

**Comment by clooudch** [comment]
> as far as I can see LDR R3, label is only possible for LDR and not for LDRB or LDRH grafik
> 
> codecoverage is not best, but will increase with branches depending on this: grafik
> 
> I don't know if it helps readability if on 16 bit instruction opcode array is set to `opcode = opcode[0]` in the onExecuteInstruction

Test will be added in the pull requests were the features are supposed to be added. So, this PR will only contain tests for data sections but not for pseudo instructions.

**Comment by clooudch** [comment]
This PR has been closed in favor of #30.

**Pull request closed without merging**